pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -136,6 +132,17 @@ func (r *ReconcileNodeMaintenance) checkLeaseSupported() error {
 // The Controller will requeue the Request to be processed again if the returned error is non-nil or
 // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
 func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+
+	// prevent reconciling the same node concurrently
+	// can happen when a lease expired and triggers a reconcile by channel source
+	var lock sync.Mutex
+	if lock, ok := reconcileLocks[request.Name]; !ok {
+		lock = &sync.Mutex{}
+		reconcileLocks[request.Name] = lock
+	}
+	lock.Lock()
+	defer lock.Unlock()
```

Comment:
at least in the logs it looked like a reconcile triggered by the channel doesn't wait until an ealier reconcile (can take a bit because of the draining) was finished, which causes CR update errors because both tried to update the CR.

Reply:
Interesting. thanks for the info

Reply:
According to [this](https://github.com/n1r1/node-maintenance-operator/blob/961e427c76cb7b194aa084475aa5395ed176aa1f/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go#L209) "reconcileHandler is never invoked concurrently with the same object".

I'm still curios to understand if reconcile can happen concurrently for different objects.

Reply:
Looks like the default is no concurrency:
https://github.com/n1r1/node-maintenance-operator/blob/961e427c76cb7b194aa084475aa5395ed176aa1f/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go#L50-L51

Reply:
hm, I see... then something else went wrong